src/zosc.c

```diff
@@ -18,11 +18,18 @@
 @end
 */
 
-#ifdef WIN32
+#include "czmq_classes.h"
+#if defined( __WINDOWS__ )
 #include <winsock2.h>           //  needed for ntohll/htonll
 #endif
-#include "czmq_classes.h"
-#if __unix__ && !__APPLE__
+#if defined(__MINGW32__)
+#define htonll(x) ((((uint64_t)htonl(x&0xFFFFFFFF)) << 32) + htonl(x >> 32))
```

Comment:
C:\projects\czmq\src\zosc.c(74): error C3861: 'htonll': identifier not found [C:\projects\czmq\build\czmq-static.vcxproj]

Reply:
weird the only thing changed was from WIN32 to __WINDOWS__ and moved after czmq_classes.h inclusion. I guess I need a Windows system to check this.

winsock2.h is also included by cmzq_prelude.h :confused:

Reply:
Ok this is not right. cmzq_prelude.h already includes winsock2.h. However it does not provide htonll. However if I include winsock2.h before cmzq_classes.h it does provide it. So somewhere in czmq something is done which results in those methods not being available

Reply:
MinGW [`winsock2.h`](https://github.com/mirror/mingw-w64/blob/a7d43b719bc6e9ee050525b4243a9ee1634555d4/mingw-w64-headers/include/winsock2.h) doesn't define `htonll`, which was the point of #2163

Reply:
I think it is caused by this line in winsock2.h

```
#if !defined(NO_EXTRA_HTON_FUNCTIONS) && !defined(__midl) && (defined(INCL_EXTRA_HTON_FUNCTIONS) || NTDDI_VERSION>=NTDDI_WIN8)
```

Reply:
Where exactly do you see this line in MinGW `winsock2.h`?

Reply:
I'm talking about winsock2.h from the Windows SDK. This line somehow returns false so the functions are never included.

https://github.com/tpn/winsdk-10/blob/9b69fd26ac0c7d0b83d378dba01080e93349c2ed/Include/10.0.10240.0/um/WinSock2.h#L1858

If your patch was correct then it should work on mingw now as well

Reply:
It might be caused by zmq.h, if you include winsock2.h before zmq.h it is fixed. I think this belongs in a separate issue. I'll fix it by including winsock2 in zosc.c before anything else like it was.